Hey Hongzhi @wehos !

I'm still getting the same error as before. 

Training the pre-trained model works without a problem:

```bash
export PIPELINE_REPO="openproblems-bio/neurips2021_multimodal_viash"
export NXF_VER=21.04.1
export PIPELINE_VERSION=1.4.0
method_id=submission_171129
task_id=predict_modality

# GENERATE PRETRAIN
pretrain_path=output/pretrain/$task_id/$method_id/pretrain.${method_id}_train.output_pretrain/

target/docker/${task_id}_methods/${method_id}_train/${method_id}_train \
  --data_dir output/datasets/$task_id \
  --output_pretrain ${pretrain_path}
```
<details><summary>Output of pre-train component</summary>

```
Copying gmt files to pretrain dir
Generating extra files
Loading pw
Generating 'pw.pkl'
Generating 'pw_multiome.pkl'
Generating 'phase2_mask.pkl'
Generating 'phase2_mask_sep.pkl'
Generating 'bf_alpha_conv4_mean_fullbatch_8000_phase2_inductive_gex2atac.pkl'
DGL backend not selected or invalid.  Assuming PyTorch for now.
Using backend: pytorch
Setting the default backend to "pytorch". You can change it in the ~/.dgl/config.json file or export the DGLBACKEND environment variable.  Valid options are: pytorch, mxnet, tensorflow (all lowercase)
Generating 'bf_alpha_conv4_mean_fullbatch_8000_phase2_inductive_gex2atac_2.pkl'
Using backend: pytorch
Generating 'bf_alpha_conv4_mean_fullbatch_8000_phase2_inductive_gex2atac_3.pkl'
Using backend: pytorch
Generating 'bf_alpha_conv4_mean_fullbatch_10000_phase2_inductive_gex2atac.pkl'
Using backend: pytorch
Generating 'bf_alpha_conv4_mean_fullbatch_10000_phase2_inductive_gex2adt_2.pkl'
Using backend: pytorch
Generating 'bf_alpha_conv4_mean_fullbatch_12000_phase2_inductive_gex2adt_sep_2.pkl'
Using backend: pytorch
Generating 'bf_alpha_conv4_mean_fullbatch_15000_phase2_inductive.pkl'
Using backend: pytorch
Generating 'f_alpha_conv4_mean_fullbatch_12000_phase2_inductive_batch_speration.pkl'
Using backend: pytorch
```
</details>

But when running the prediction component, I get the following error:

```bash
# CITE GEX2ADT
dataset_id=openproblems_bmmc_cite_phase2_rna
dataset_path=output/datasets/$task_id/$dataset_id/$dataset_id.censor_dataset
pred_path=output/predictions/$task_id/$dataset_id/$dataset_id

target/docker/${task_id}_methods/${method_id}/${method_id} \
  --input_train_mod1 ${dataset_path}.output_train_mod1.h5ad \
  --input_train_mod2 ${dataset_path}.output_train_mod2.h5ad \
  --input_test_mod1 ${dataset_path}.output_test_mod1.h5ad \
  --input_pretrain ${pretrain_path} \
  --output ${pred_path}.${method_id}.output.h5ad
```
Output of prediction component:
```
INFO:root:Reading `h5ad` files...
Setting the default backend to "pytorch". You can change it in the ~/.dgl/config.json file or export the DGLBACKEND environment variable.  Valid options are: pytorch, mxnet, tensorflow (all lowercase)
DGL backend not selected or invalid.  Assuming PyTorch for now.
Using backend: pytorch
Traceback (most recent call last):
  File "/viash_automount/home/ec2-user/ssd/tmp//viash-run-submission_171129-9m9iTP", line 93, in <module>
    y_pred.append(evaluate(model, args).numpy())
  File "/viash_automount/home/ec2-user/ssd/tmp//viash-run-submission_171129-9m9iTP", line 75, in evaluate
    logits = mod(g, bf, args)
  File "/usr/local/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
TypeError: forward() takes 3 positional arguments but 4 were given
```

I can (again) send you the output of the pretrain component so that you can try to get the prediction component fixed. Let me know if you want me to send these files to you.

I think the issue, in large, is caused by the two components having two totally different definitions of WeightedGCN4 ([run](https://github.com/openproblems-bio/neurips2021_multimodal_topmethods/blob/main/src/predict_modality/methods/DANCE/resources/graph_util.py#L152) vs [train](https://github.com/openproblems-bio/neurips2021_multimodal_topmethods/blob/main/src/predict_modality/methods/DANCE/train/hetero_arg_version_v5.py#L483)). I suggest looking at the [code by team AXX](https://github.com/openproblems-bio/neurips2021_multimodal_topmethods/tree/main/src/predict_modality/methods/AXX); they added most of the code to the resources directory to ensure that they're working with the same helper functions in both the train and run components.

Please get this fixed as soon as possible. Make sure you're able to run the whole pipeline (`src/predict_modality/methods/DANCE/test.sh`).

--- a/src/predict_modality/methods/DANCE/resources/graph_util.py
+++ b/src/predict_modality/methods/DANCE/resources/graph_util.py
@@ -150,112 +150,246 @@ def graph_construction(meta, train_mod1, train_mod2, test_mod1, pretrain_path, b
     return g, batch_features
     
 class WeightedGCN4(nn.Module):
-    def __init__(self, FEATURE_SIZE, hid_feats, out_feats, bf_size, args):
+    def __init__(self, hid_feats, out_feats):
         super().__init__()
+        self.opw = args.only_pathway
+        self.npw = args.no_pathway
+        self.nrc = args.no_readout_concatenate
         
-        self.extra_encoder = nn.Linear(bf_size, hid_feats)
-        self.args = args
-
-        self.embed_cell = nn.Embedding(2, hid_feats)
+        if batch_features is not None:
+            self.extra_encoder = nn.Linear(batch_features.shape[1], hid_feats)
+        if args.overlap:
+            self.ov_encoder = nn.Linear(2, hid_feats)
+            
+        if args.cell_init == 'none':
+            self.embed_cell = nn.Embedding(2, hid_feats)
+        else:
+            self.embed_cell = nn.Linear(100, hid_feats)
+            
         self.embed_feat = nn.Embedding(FEATURE_SIZE, hid_feats)
-
+        
         self.input_linears = nn.ModuleList()
         self.input_acts = nn.ModuleList()
         self.input_norm = nn.ModuleList()
-        for i in range(2):
+        for i in range((args.embedding_layers-1)*2):
             self.input_linears.append(nn.Linear(hid_feats, hid_feats))
-        if self.args['activation'] == 'gelu':
-            for i in range(2):
+        if args.activation == 'gelu':
+            for i in range((args.embedding_layers-1)*2):
                 self.input_acts.append(nn.GELU())
-        elif self.args['activation'] == 'leaky_relu':
-            for i in range(2):
+        elif args.activation == 'prelu':
+            for i in range((args.embedding_layers-1)*2):
+                self.input_acts.append(nn.PReLU())
+        elif args.activation == 'relu':
+            for i in range((args.embedding_layers-1)*2):
+                self.input_acts.append(nn.ReLU())
+        elif args.activation == 'leaky_relu':
+            for i in range((args.embedding_layers-1)*2):
                 self.input_acts.append(nn.LeakyReLU()) 
-
-        for i in range(2):
-            self.input_norm.append(nn.GroupNorm(4, hid_feats))
-
-        self.edges = ['entail', 'occur', 'pathway']
-
+        if args.normalization == 'batch':
+            for i in range((args.embedding_layers-1)*2):
+                self.input_norm.append(nn.BatchNorm1d(hid_feats))
+        elif args.normalization == 'layer':
+            for i in range((args.embedding_layers-1)*2):
+                self.input_norm.append(nn.LayerNorm(hid_feats))
+        elif args.normalization == 'group':
+            for i in range((args.embedding_layers-1)*2):
+                self.input_norm.append(nn.GroupNorm(4, hid_feats))
+        
+        if self.opw:
+            self.edges = ['entail', 'pathway']
+        elif self.npw:
+            self.edges = ['entail', 'occur']
+        else:
+            self.edges = ['entail', 'occur', 'pathway']
+        
         self.conv_layers = nn.ModuleList()
+        if args.residual == 'res_cat':
+            self.conv_layers.append(dglnn.HeteroGraphConv(dict(zip(self.edges, [dglnn.SAGEConv(in_feats=hid_feats, out_feats=hid_feats, aggregator_type=args.agg_function, norm=None) for i in range(len(self.edges))])), aggregate='stack'))
+            for i in range(args.conv_layers-1):
+                self.conv_layers.append(dglnn.HeteroGraphConv(dict(zip(self.edges, [dglnn.SAGEConv(in_feats=hid_feats*2, out_feats=hid_feats, aggregator_type=args.agg_function, norm=None) for i in range(len(self.edges))])), aggregate='stack'))
 
-        self.conv_layers.append(dglnn.HeteroGraphConv(dict(zip(self.edges, [dglnn.SAGEConv(in_feats=hid_feats, out_feats=hid_feats, aggregator_type=self.args['agg_function'], norm=None) for i in range(len(self.edges))])), aggregate='stack'))
-        for i in range(self.args['conv_layers']-1):
-            self.conv_layers.append(dglnn.HeteroGraphConv(dict(zip(self.edges, [dglnn.SAGEConv(in_feats=hid_feats*2, out_feats=hid_feats, aggregator_type=self.args['agg_function'], norm=None) for i in range(len(self.edges))])), aggregate='stack'))
-
+        else:
+            for i in range(args.conv_layers):
+                self.conv_layers.append(dglnn.HeteroGraphConv(dict(zip(self.edges, [dglnn.SAGEConv(in_feats=hid_feats, out_feats=hid_feats, aggregator_type=args.agg_function, norm=None) for i in range(len(self.edges))])), aggregate='stack'))    
+        
         self.conv_acts = nn.ModuleList()
         self.conv_norm = nn.ModuleList()
-        if self.args['activation'] == 'gelu':
-            for i in range(self.args['conv_layers']*2):
+        if args.activation == 'gelu':
+            for i in range(args.conv_layers*2):
                 self.conv_acts.append(nn.GELU())
-        elif self.args['activation'] == 'leaky_relu':
-            for i in range(self.args['conv_layers']*2):
+        elif args.activation == 'prelu':
+            for i in range(args.conv_layers*2):
+                self.conv_acts.append(nn.PReLU())
+        elif args.activation == 'relu':
+            for i in range(args.conv_layers*2):
+                self.conv_acts.append(nn.ReLU())
+        elif args.activation == 'leaky_relu':
+            for i in range(args.conv_layers*2):
                 self.conv_acts.append(nn.LeakyReLU())
-
-        for i in range(self.args['conv_layers']*len(self.edges)):
-            self.conv_norm.append(nn.GroupNorm(4, hid_feats))
-
+            
+        if args.normalization == 'batch':
+            for i in range(args.conv_layers*len(self.edges)):
+                self.conv_norm.append(nn.BatchNorm1d(hid_feats))
+        elif args.normalization == 'layer':
+            for i in range(args.conv_layers*len(self.edges)):
+                self.conv_norm.append(nn.LayerNorm(hid_feats))
+        elif args.normalization == 'group':
+            for i in range(args.conv_layers*len(self.edges)):
+                self.conv_norm.append(nn.GroupNorm(4, hid_feats))
+        
         self.att_linears = nn.ModuleList()
+        if args.pathway_aggregation == 'attention':
+            for i in range(args.conv_layers):
+                self.att_linears.append(nn.Linear(hid_feats, hid_feats))
+        elif args.pathway_aggregation == 'one_gate':
+            for i in range(args.conv_layers):
+                self.att_linears.append(nn.Linear(hid_feats*3, hid_feats))
+        elif args.pathway_aggregation == 'two_gate': 
+            for i in range(args.conv_layers*2):
+                self.att_linears.append(nn.Linear(hid_feats*2, hid_feats))
+        elif args.pathway_aggregation == 'cat':
+            for i in range(args.conv_layers):
+                self.att_linears.append(nn.Linear(hid_feats*2, hid_feats))
+                
         self.readout_linears = nn.ModuleList()
         self.readout_acts = nn.ModuleList()
-        self.readout_linears.append(nn.Linear(hid_feats*self.args['conv_layers'], out_feats))
-
-        if self.args['activation'] == 'gelu':
-            for i in range(self.args['readout_layers']-1):
+        if self.nrc:
+            for i in range(args.readout_layers-1):
+                self.readout_linears.append(nn.Linear(hid_feats, hid_feats))
+            self.readout_linears.append(nn.Linear(hid_feats, out_feats))
+        else:
+            for i in range(args.readout_layers-1):
+                self.readout_linears.append(nn.Linear(hid_feats*args.conv_layers, hid_feats*args.conv_layers))
+            self.readout_linears.append(nn.Linear(hid_feats*args.conv_layers, out_feats))
+            
+        if args.activation == 'gelu':
+            for i in range(args.readout_layers-1):
                 self.readout_acts.append(nn.GELU())
-        elif self.args['activation'] == 'leaky_relu':
-            for i in range(self.args['readout_layers']-1):
+        elif args.activation == 'prelu':
+            for i in range(args.readout_layers-1):
+                self.readout_acts.append(nn.PReLU())
+        elif args.activation == 'relu':
+            for i in range(args.readout_layers-1):
+                self.readout_acts.append(nn.ReLU())
+        elif args.activation == 'leaky_relu':
+            for i in range(args.readout_layers-1):
                 self.readout_acts.append(nn.LeakyReLU())
 
-    def attention_agg(self, layer, h0, h):
+    def attention_agg(self, layer, h0, h, args):
         # h: h^{l-1}, dimension: (batch, hidden)
         # feats: result from two conv(cell conv and pathway conv), stacked together; dimension: (batch, 2, hidden)
         if h.shape[1]==1:
             return self.conv_norm[layer*len(self.edges)+1](h.squeeze(1))
+        elif args.pathway_aggregation == 'sum':
+            return h[:, 0, :] + h[:, 1, :]
         else:
-            h1 = self.conv_norm[layer*len(self.edges)+1](h[:, 0, :])
-            h2 = self.conv_norm[layer*len(self.edges)+2](h[:, 1, :])
-
-        return (1-self.args['pathway_alpha']) * h1 + self.args['pathway_alpha'] * h2
-
-    def conv(self, graph, layer, h, hist):
+            h1 = h[:, 0, :]
+            h2 = h[:, 1, :]
+            
+            if args.subpath_activation:
+                h1 = F.leaky_relu(h1)
+                h2 = F.leaky_relu(h2)
+                
+            h1 = self.conv_norm[layer*len(self.edges)+1](h1)
+            h2 = self.conv_norm[layer*len(self.edges)+2](h2)
+        
+        if args.pathway_aggregation == 'attention':
+            feats = torch.stack([h1, h2], 1)
+            att = torch.transpose(F.softmax(torch.matmul(feats, self.att_linears[layer](h0).unsqueeze(-1)), 1), 1, 2)
+            feats = torch.matmul(att, feats)
+            return feats.squeeze(1)
+        elif args.pathway_aggregation == 'one_gate':
+            att = torch.sigmoid(self.att_linears[layer](torch.cat([h0, h1, h2], 1)))
+            return att * h1 + (1-att) * h2
+        elif args.pathway_aggregation == 'two_gate':
+            att1 = torch.sigmoid(self.att_linears[layer*2](torch.cat([h0, h1], 1)))
+            att2 = torch.sigmoid(self.att_linears[layer*2+1](torch.cat([h0, h2], 1)))
+            return att1 * h1 + att2 * h2
+        elif args.pathway_aggregation == 'alpha':
+            return (1-args.pathway_alpha) * h1 + args.pathway_alpha * h2
+        elif args.pathway_aggregation == 'cat':
+            return self.att_linears[layer](torch.cat([h1, h2], 1))
+        
+    def conv(self, graph, layer, h, hist, args):
         h0 = hist[-1]
-        h = self.conv_layers[layer](graph, h, mod_kwargs=dict(zip(self.edges, [{'edge_weight':  F.dropout(graph.edges[self.edges[i]].data['weight'], p=self.args['edge_dropout'], training=self.training)} for i in range(len(self.edges))])))
-
-        h = {'feature': F.dropout(self.conv_acts[layer*2](self.attention_agg(layer, h0['feature'], h['feature'])), p=self.args['model_dropout'], training=self.training),
-         'cell': F.dropout(self.conv_acts[layer*2+1](self.conv_norm[layer*len(self.edges)](h['cell'].squeeze(1))), p=self.args['model_dropout'], training=self.training)}
+        h = self.conv_layers[layer](graph, h, mod_kwargs=dict(zip(self.edges, [{'edge_weight':  F.dropout(graph.edges[self.edges[i]].data['weight'], p=args.edge_dropout, training=self.training)} for i in range(len(self.edges))])))
+        
+        if args.model_dropout>0:
+            h = {'feature': F.dropout(self.conv_acts[layer*2](self.attention_agg(layer, h0['feature'], h['feature'], args)), p=args.model_dropout, training=self.training),
+             'cell': F.dropout(self.conv_acts[layer*2+1](self.conv_norm[layer*len(self.edges)](h['cell'].squeeze(1))), p=args.model_dropout, training=self.training)}
+        else:
+            h = {'feature': self.conv_acts[layer*2](self.attention_agg(layer, h0['feature'], h['feature'], args)),
+             'cell': self.conv_acts[layer*2+1](self.conv_norm[layer*len(self.edges)](h['cell'].squeeze(1)))}
+            
         return h
-
-    def forward(self, graph, batch_features):
+        
+    def forward(self, graph, batch_features, args):
         input1 = F.leaky_relu(self.embed_feat(graph.nodes['feature'].data['id']))
         input2 = F.leaky_relu(self.embed_cell(graph.nodes['cell'].data['id']))
-        input2 += F.leaky_relu(F.dropout(self.extra_encoder(batch_features), p=0.2, training=self.training))[:input2.shape[0]]
-
+        
+        if not args.no_batch_features:
+            input2 += F.leaky_relu(F.dropout(self.extra_encoder(batch_features), p=0.2, training=self.training))[:input2.shape[0]]
+        
+        if args.overlap:
+            input1 += F.leaky_relu(self.ov_encoder(gex_feature))
+            
         hfeat = input1
         hcell = input2
-        for i in range(1, 2):
+        for i in range(args.embedding_layers-1, (args.embedding_layers-1)*2):
             hfeat = self.input_linears[i](hfeat)
             hfeat = self.input_acts[i](hfeat)
-            hfeat = self.input_norm[i](hfeat)
-            hfeat = F.dropout(hfeat, p=self.args['model_dropout'], training = self.training)
-
-        for i in range(1):
+            if args.normalization != 'none':
+                hfeat = self.input_norm[i](hfeat)
+            if args.model_dropout>0:
+                hfeat = F.dropout(hfeat, p=args.model_dropout, training = self.training)
+                
+        for i in range(args.embedding_layers-1):
             hcell = self.input_linears[i](hcell)
             hcell = self.input_acts[i](hcell)
-            hcell = self.input_norm[i](hcell)
-            hcell = F.dropout(hcell, p=self.args['model_dropout'], training = self.training)
+            if args.normalization != 'none':
+                hcell = self.input_norm[i](hcell)
+            if args.model_dropout>0:
+                hcell = F.dropout(hcell, p=args.model_dropout, training = self.training)
 
         h = {'feature': hfeat, 'cell': hcell}
         hist = [h]
+        
+        for i in range(args.conv_layers):
+            if i==0 or args.residual=='none':
+                pass
+            elif args.residual == 'res_add':
+                if args.initial_residual:
+                    h = {'feature': h['feature']+hist[0]['feature'],
+                    'cell': h['cell']+hist[0]['cell']}
 
-        for i in range(self.args['conv_layers']):
-            if i>0:
-                h = {'feature': torch.cat([h['feature'], hist[0]['feature']], 1),
+                else:
+                    h = {'feature': h['feature']+hist[-2]['feature'],
+                    'cell': h['cell']+hist[-2]['cell']}
+
+            elif args.residual == 'res_cat':
+                if args.initial_residual:
+                    h = {'feature': torch.cat([h['feature'], hist[0]['feature']], 1),
                     'cell': torch.cat([h['cell'], hist[0]['cell']], 1)}
+                else:
+                    h = {'feature': torch.cat([h['feature'], hist[-2]['feature']], 1),
+                    'cell': torch.cat([h['cell'], hist[-2]['cell']], 1)}
 
-            h = self.conv(graph, i, h, hist)
+            h = self.conv(graph, i, h, hist, args)
             hist.append(h)
-
-        h = torch.cat([i['cell'] for i in hist[1:]], 1)
+        
+        if not self.nrc:
+            h = torch.cat([i['cell'] for i in hist[1:]], 1)
+        else:
+            h = h['cell']
+            
+        for i in range(args.readout_layers-1):
+            h = self.readout_linears[i](h)
+            h = F.dropout(readout_acts[i](h), p=args.model_dropout, training = self.training)
         h = self.readout_linears[-1](h)
-
+        
+        if args.output_relu == 'relu':
+            return F.relu(h)
+        elif args.output_relu == 'leaky_relu':
+            return F.leaky_relu(h)
+        
         return h

